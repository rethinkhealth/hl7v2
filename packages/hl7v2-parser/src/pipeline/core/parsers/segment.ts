import type { HL7v2Node } from '@rethinkhealth/hl7v2-ast';
import type { ISegmentParser, SegmentParseInput } from '../../interfaces';
import { splitByString } from '../../utils';
import { FieldParser } from './field';

/**
 * Default segment parser for standard HL7v2 segments
 */
export class SegmentParser implements ISegmentParser {
  segmentType = 'DEFAULT';
  private fieldParser = new FieldParser();

  canParse(input: SegmentParseInput): boolean {
    return input.type === 'segment' && input.text.trim().length > 0;
  }

  parse(input: SegmentParseInput): HL7v2Node | null {
    if (!this.canParse(input)) {
      return null;
    }

    const fields = splitByString(input.text, input.context.delimiters.field);
    if (!fields[0] || fields.length === 0) {
      return null;
    }

    const segmentNode: HL7v2Node = {
      type: 'segment',
      name: fields[0].value,
      index: input.index,
      delimiter: input.context.delimiters.field,
      children: [],
      position: {
        start: { line: input.line, column: 1, offset: input.start },
        end: {
          line: input.line,
          column: input.text.length + 1,
          offset: input.end,
        },
      },
    };

    segmentNode.children = this.parseFields(fields, input);
    return segmentNode;
  }

  private parseFields(
    fields: Array<{ value: string; start: number; end: number }>,
    input: SegmentParseInput
  ): HL7v2Node[] {
    const children: HL7v2Node[] = [];

    for (let i = 0; i < fields.length; i++) {
      const field = fields[i];
      if (!field) {
        continue;
      }

      const fieldNode = this.fieldParser.parse({
        type: 'field',
        text: field.value,
        start: input.start + field.start,
        end: input.start + field.end,
        index: i,
        fieldIndex: i,
        line: input.line,
        column: field.start + 1,
        context: input.context,
      });

      if (fieldNode) {
        children.push(fieldNode);
      }
    }

    return children;
  }
}
