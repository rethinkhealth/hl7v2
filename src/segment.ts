import { SEQUENCE_STARTING_INDEX } from "./constants";
import { DefaultDelimiters, IDelimiters } from "./delimiters";
import { Element, IElement } from "./element";
import { SegmentType } from "./enum";

export interface ISegment {
  delimiters: IDelimiters;
  fields: IElement[];
  name: SegmentType;
  raw: string;
  toJson(): any;
}

export interface SegmentOptions {
  delimiters: IDelimiters;
}

export abstract class SegmentBase implements ISegment {
  public readonly raw: string;

  private _delimiters: IDelimiters;
  public get delimiters(): IDelimiters {
    return this._delimiters;
  }

  private _name: SegmentType;
  public get name(): SegmentType {
    return this._name;
  }

  abstract fields: IElement[];

  constructor(segment: string, private options?: SegmentOptions) {
    this.raw = segment;
    this._name = "" as any;
    this._delimiters = {} as any;

    // Configuration steps
    this.setupDelimiters();
    this.setupName();
  }

  public toJson() {
    const response = {} as any;
    this.fields.forEach((field) => {
      response[field.sequence] = field.toJson();
    });
    return response;
  }

  private setupDelimiters() {
    this._delimiters = this.options?.delimiters || DefaultDelimiters;
  }

  private setupName() {
    const name = this.raw.split(this.delimiters.fieldSeparator)[0];
    if (!Object.values(SegmentType).includes(name as unknown as SegmentType)) {
      // TODO: Make it custom error handling for better management of errors.
      throw new Error(
        `Segment name not recognized. ${name} is not in the list.`
      );
    }
    this._name = SegmentType[name as keyof typeof SegmentType];
  }
}

export class Segment extends SegmentBase {
  private _fields: IElement[];
  public get fields(): IElement[] {
    return this._fields;
  }

  constructor(segment: string, options?: SegmentOptions) {
    super(segment, options);
    this._fields = [];

    this.setupElements();
  }

  private setupElements() {
    const fields = this.raw.split(this.delimiters.fieldSeparator);
    // Create the elements by running through the elements
    fields
      // ignore the first element because it is the name of the segment
      .slice(1)
      // run through the elements
      .forEach((value, index) => {
        const sequence = index + SEQUENCE_STARTING_INDEX;
        const element = new Element(value, sequence.toString(), {
          delimiters: this.delimiters,
        });
        this.fields.push(element);
      });
  }
}
