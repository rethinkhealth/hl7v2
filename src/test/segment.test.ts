import { Segment } from "../segment";

describe("Segment", () => {
  it("should have a line if provided", () => {
    const rawMessage =
      "MSH|^~\\&|Ntierprise|Ntierprise Clinic|Healthmatics EHR";
    const segment = new Segment(rawMessage, { line: 1 });
    expect(segment.line).toEqual(1);
  });

  it("should have line undefined by default", () => {
    const rawMessage =
      "MSH|^~\\&|Ntierprise|Ntierprise Clinic|Healthmatics EHR";
    const segment = new Segment(rawMessage);
    expect(segment.line).toEqual(undefined);
  });
});

describe.each([
  ["AIG", "AIG|1||PULLEN^Pullen, Jeri|P^^CSI"],
  [
    "PID",
    "PID|1||150||Bond^James^^007||19770920|M|||007 Soho Lane^^Cary^NC^27511||(919)007-0007^^^^^919^0070007~(777)707-0707^^CP^^^777^7070707~^NET^X.400^007@BritishSecretService.com|(919)851-6177 X007^^^^^919^8516177^007||M||150|007-00-0007|||||||||||N",
  ],
  [
    "MSH",
    "MSH|^~&|Ntierprise|Ntierprise Clinic|Healthmatics EHR|Healthmatics Clinic|20190423114154||SIU^S12|8907-45|P|2.3|||NE|NE",
  ],
])("HL7v2 Segment for %s", (expectedName, input) => {
  it("should have the raw original segment", () => {
    // When
    const segment = new Segment(input);

    // Then
    expect(segment.raw).toEqual(input);
  });

  it("Name for the message should be %s", () => {
    // When
    const segment = new Segment(input);

    // Then
    expect(segment.name).toEqual(expectedName);
  });

  it("should raise an exception if the Segment Name is not identified", () => {
    // Given
    const incompatibleName = "XXR";
    const modifiedInput = input.replace(expectedName, incompatibleName);

    // When
    expect(() => {
      new Segment(modifiedInput);
    }).toThrowError(
      `Segment name not recognized. ${incompatibleName} is not in the list.`
    );
  });
});

describe("PID message", () => {
  it("should return the elements with sequence for PID", () => {
    // Given
    const input =
      "PID|1||150||Bond^James^^007||19770920|M|||007 Soho Lane^^Cary^NC^27511||(919)007-0007^^^^^919^0070007~(777)707-0707^^CP^^^777^7070707~^NET^X.400^007@BritishSecretService.com|(919)851-6177 X007^^^^^919^8516177^007||M||150|007-00-0007|||||||||||N";

    // When
    const segment = new Segment(input);

    // Then
    expect(segment.fields.length).toBe(30);
    expect(segment.fields[0].sequence).toEqual("1");
    expect(segment.fields[0].value).toEqual("1");
  });

  it("should return JSON format", () => {
    // Given
    const input =
      "PID|1||150||Somai^Melek^^007||19770920|M|||007 Soho Lane^^Cary^NC^27511||(919)007-0007^^^^^919^0070007~(777)707-0707^^CP^^^777^7070707~^NET^X.400^007@BritishSecretService.com|(919)851-6177 X007^^^^^919^8516177^007||M||150|007-00-0007|||||||||||N";

    // When
    const segment = new Segment(input);

    // Then
    expect(segment.toJson()).toMatchSnapshot();
  });
});

describe("MSH message", () => {
  it("should return the elements with sequence for MSH", () => {
    // Given
    const input =
      "MSH|^~&|Ntierprise|Ntierprise Clinic|Healthmatics EHR|Healthmatics Clinic|20190423114154||SIU^S12|8907-45|P|2.3|||NE|NE";

    // When
    const segment = new Segment(input);

    // Then
    expect(segment.fields.length).toBe(15);
  });
});
