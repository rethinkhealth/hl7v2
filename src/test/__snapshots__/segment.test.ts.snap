// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PID message should return JSON format 1`] = `
{
  "1": "1",
  "10": "",
  "11": {
    "1": "007 Soho Lane",
    "2": "",
    "3": "Cary",
    "4": "NC",
    "5": "27511",
  },
  "12": "",
  "13": [
    {
      "1": "(919)007-0007",
      "2": "",
      "3": "",
      "4": "",
      "5": "",
      "6": "919",
      "7": "0070007",
    },
    {
      "1": "(777)707-0707",
      "2": "",
      "3": "CP",
      "4": "",
      "5": "",
      "6": "777",
      "7": "7070707",
    },
    {
      "1": "",
      "2": "NET",
      "3": "X.400",
      "4": "007@BritishSecretService.com",
    },
  ],
  "14": {
    "1": "(919)851-6177 X007",
    "2": "",
    "3": "",
    "4": "",
    "5": "",
    "6": "919",
    "7": "8516177",
    "8": "007",
  },
  "15": "",
  "16": "M",
  "17": "",
  "18": "150",
  "19": "007-00-0007",
  "2": "",
  "20": "",
  "21": "",
  "22": "",
  "23": "",
  "24": "",
  "25": "",
  "26": "",
  "27": "",
  "28": "",
  "29": "",
  "3": "150",
  "30": "N",
  "4": "",
  "5": {
    "1": "Somai",
    "2": "Melek",
    "3": "",
    "4": "007",
  },
  "6": "",
  "7": "19770920",
  "8": "M",
  "9": "",
}
`;
